File: services/entity-resolution-service/src/fusion_base/domain/feedback.rs
use chrono::Utc;

use crate::models::cluster::{ResolvedCluster, ReviewQueueItem, SubmitReviewRequest};

pub fn apply_review(
    cluster: &ResolvedCluster,
    review_item: Option<&ReviewQueueItem>,
    request: &SubmitReviewRequest,
) -> (ResolvedCluster, Option<ReviewQueueItem>) {
    let now = Utc::now();
    let mut updated_cluster = cluster.clone();
    let mut updated_review = review_item.cloned();

    match request.decision.as_str() {
        "confirm_match" => {
            updated_cluster.status = "resolved".to_string();
            updated_cluster.requires_review = false;
        }
        "split_cluster" => {
            updated_cluster.status = "split_requested".to_string();
            updated_cluster.requires_review = false;
        }
        "reject_match" => {
            updated_cluster.status = "rejected".to_string();
            updated_cluster.requires_review = false;
        }
        _ => {
            updated_cluster.status = "manually_resolved".to_string();
            updated_cluster.requires_review = false;
        }
    }
    updated_cluster.updated_at = now;

    if let Some(review_item) = &mut updated_review {
        review_item.status = match request.decision.as_str() {
            "split_cluster" => "split_requested",
            "reject_match" => "rejected",
            _ => "resolved",
        }
        .to_string();
        review_item.reviewed_by = request.reviewed_by.clone();
        review_item.notes = request.notes.clone().unwrap_or_default();
        review_item.updated_at = now;
    }

    (updated_cluster, updated_review)
}

File: services/entity-resolution-service/internal/domain/feedback.go
// Package domain ports fusion_base/domain — feedback (apply_review),
// deduplication (synthesize_entity_records), merge (synthesize_golden_record).
package domain

import (
	"time"

	"github.com/openfoundry/openfoundry-go/services/entity-resolution-service/internal/models"
)

// ApplyReview ports `feedback::apply_review`.
//
// Returns updated cluster + optional updated review item. The original
// inputs are not mutated.
func ApplyReview(cluster models.ResolvedCluster, reviewItem *models.ReviewQueueItem, request models.SubmitReviewRequest) (models.ResolvedCluster, *models.ReviewQueueItem) {
	now := time.Now().UTC()

	updated := cluster
	switch request.Decision {
	case "confirm_match":
		updated.Status = "resolved"
		updated.RequiresReview = false
	case "split_cluster":
		updated.Status = "split_requested"
		updated.RequiresReview = false
	case "reject_match":
		updated.Status = "rejected"
		updated.RequiresReview = false
	default:
		updated.Status = "manually_resolved"
		updated.RequiresReview = false
	}
	updated.UpdatedAt = now

	if reviewItem == nil {
		return updated, nil
	}
	updatedReview := *reviewItem
	switch request.Decision {
	case "split_cluster":
		updatedReview.Status = "split_requested"
	case "reject_match":
		updatedReview.Status = "rejected"
	default:
		updatedReview.Status = "resolved"
	}
	updatedReview.ReviewedBy = request.ReviewedBy
	notes := ""
	if request.Notes != nil {
		notes = *request.Notes
	}
	updatedReview.Notes = notes
	updatedReview.UpdatedAt = now

	return updated, &updatedReview
}
